Date range validation

shahi645 opened the pull request:
### What is the context of this PR?
- Added date range validation for a period between to dates, allows for a minimum limit or/and a maximum limit to be set 
- Added period validation given a referenced date for single dates
- Runtime validation for invalid schemas with date ranges

### How to review 
- open `test_date_range_period_validation.json` and test out the question level validation
- also open `test_single_date_period_validation.json` and test out single date period validation
- Schema-validator PR: https://github.com/ONSdigital/eq-schema-validator/pull/44

dwyeradam commented on data/en/test_date_range_period_validation.json:
```diff
@@ -0,0 +1,53 @@
+                            "title": "Enter Date Range",
+                            "type": "DateRange",
+                            "period_limits": {
+                              "min_value": 23,
```
23 what? Years, Months, Days, Minutes, Seconds, Flamingos? Check the format of offset on date routing.

dwyeradam replied:
Also does min_value define what this is? Is min_range clearer?

shahi645 replied:
its days, and min_limit perhaps ?

shahi645 replied:
ok so i think `minimum_day_limit` will clear it up a little ?

dwyeradam replied:
My point is that we shouldn't limit ourselves to just days. Years, Months, Weeks and Days could all be valid requirements in the future.

shahi645 replied:
This is a bit like the average stuff in the calculated validation, should i add the ability for it when there are no business requirements for it? And then a few days down the line im backing the changes out because its not needed. The card only asks for period limits to be set in days. I am more than happy to add the ability for other units of time but if it is reuested

dwyeradam replied:
Consider using dateutil.relativedelta in your date comparison to help compare difference in years, months, days etc.

dwyeradam commented on app/forms/questionnaire_form.py:
```diff
@@ -39,9 +39,13 @@ def validate(self):
                 period_to_id = question['answers'][1]['id']
+                messages = None
+                if 'validation' in question:
+                    messages = question['validation'].get('messages')
```
Good addition.

dwyeradam commented on app/forms/questionnaire_form.py:
```diff
@@ -104,10 +104,10 @@ def _validate_calculated_question(self, question, target_answer, target_total):
-                maximum = limits['max_value']
+            if 'minimum_days_limit' in limits:
+                minimum = limits['minimum_days_limit']
+            if 'maximum_days_limit' in limits:
```
This hasn't addressed my previous comment. "We shouldn't limit ourselves to just days. Years, Months, Weeks and Days could all be valid requirements in the future." This can quite easily be done by having a structure as below and using **dateutil.relativedelta**.
```
"min": {
   "days": 23
}

```
or 

```
"min": {
   "months": 3
   "years": 1
}
```

shahi645 replied:
@dwyeradam I did reply on your last change. There is no requirements for month and year in between a date range. I have double checked with Donna also everything upcoming is for days in between, the trello card also describes days as the only requirement. Also months/years can be representing in days also. 
In regard the `relativedelta` I am just in process of changing over but its causing some issues so still working througth that

jonnyshaw89 replied:
@shahi645 it may not be required now, but if it is in the future then we are going to end up having a messy schema.
This pattern has already been introduced so it seems sensible to me to keep this kind of definition consistent across the schema.

dwyeradam replied:
@shahi645. I agree with Jon. We're building a feature not a survey. That feature would be better written allowing for more than just days. Also the code required for that does not add any additional complication. As I keep saying if you research the use of relativedelta you'd realise days, months, years, weeks can all be done in a single line.

shahi645 replied:
Yes i don't disagree with that i have looked in to relativedelta its simple enough im working on those changes currently. In regards to the days/weeks/months there is no future requirement for it. This is the exact same thing that was said for average on the calculated question type and it just got backed out in the end.

MIslam92 commented on app/helpers/form_helper.py:
```diff
@@ -11,14 +11,15 @@
     """
     Returns the form necessary for the location given a get request, plus any template arguments
 
+    :param schema: schema
```
Might want to expand on this docstring i.e. the current schema json or something

MIslam92 commented on app/forms/date_form.py:
```diff
@@ -84,3 +98,61 @@ class MonthYearDateForm(Form):
+def get_referenced_offset_value(answer_min_or_max, answer_store, metadata):
+    """
+    Gets value of the referenced date type, whether it is a value,
+    id of an answer or a meta date. Then adds/subtracts offset from that value and
```
Did you have a stroke midway and forget to finish this off?

shahi645 replied:
🌚

MIslam92 commented on app/validation/error_messages.py:
```diff
@@ -22,4 +22,8 @@
     'INVALID_DATE': 'Enter a valid date.',
     'INVALID_DATE_RANGE': "Enter a 'period to' date later than the 'period from' date.",
+    'DATE_PERIOD_TOO_SMALL': 'Enter a reporting period greater than %(min)s.',
+    'DATE_PERIOD_TOO_BIG': 'Enter a reporting period less than %(max)s.',
```
Might want to use `DATE_PERIOD_TOO_LARGE` as this is the terminology we use in runner

nealedj commented on app/forms/date_form.py:
```diff
@@ -41,6 +48,13 @@ class DateForm(Form):
+        if 'validation' in answer:
+            messages = answer['validation'].get('messages')
+        minimum_date, maximum_date = get_min_max_dates_for_single_date_period_validation(answer, answer_store, metadata)
+        validate_with += [SingleDatePeriodCheck(messages=messages, minimum_date=minimum_date, maximum_date=maximum_date)]
```
Minor point - but let's not create a list just to append a single item to a list. Use `validate_with.append()`

Whilst there - maybe also change the same thing on line 49.

nealedj commented on app/forms/date_form.py:
```diff
@@ -84,3 +98,62 @@ class MonthYearDateForm(Form):
+    if 'maximum' in answer:
+        maximum_referenced_date = get_referenced_offset_value(answer['maximum'], answer_store, metadata)
+
+    if minimum_referenced_date and maximum_referenced_date:
```
Could you please add a little comment to mention that this should never happen as validator should prevent it.

nealedj commented on app/forms/date_form.py:
```diff
@@ -84,3 +98,62 @@ class MonthYearDateForm(Form):
+    elif 'meta' in answer_min_or_max:
+        value = get_metadata_value(metadata, answer_min_or_max['meta'])
+    elif 'answer_id' in answer_min_or_max:
+        value = get_answer_store_value(answer_min_or_max['answer_id'], answer_store, group_instance=0)
```
This would fail in a repeating group wouldn't it? As it would always validate against the value of the answer in the first group instance, not the current group instance.

shahi645 replied:
Ticket raised in planning backlog to address this, ass per Davids request

dwyeradam replied:
@nealedj, @shahi645. Why are we introducing tech debt? In my opinion if we know this feature will fail in certain conditions it shouldn't be put in. I wouldn't class this as an edge case as we already know we're likely to need this in repeating groups for both Census and LFS.

nealedj replied:
I agree that it's not an edge case and that it's not good to introduce tech debt.

But on balance I don't think we gain anything by holding up this feature before fixing it given the fact that it's not a bug that will manifest itself in any current questionnaire.

It's also a fix that needs to address number range validation under the same conditions.

dwyeradam replied:
Having spoken with @nealedj, I realise this is the same format as used by numbers so a separate PR will be needed for that either way. To get this PR through then i'm happy for this to be addressed at the same time and in the same fashion as will be needed for numbers.

nealedj commented on app/forms/date_form.py:
```diff
@@ -84,3 +98,62 @@ class MonthYearDateForm(Form):
+                                                           months=offset.get('months', 0),
+                                                           days=offset.get('days', 0))
+    else:
+        value = convert_to_datetime(value)
```
`value = convert_to_datetime(value)` is repeated. This could be simplified to:

```
value = convert_to_datetime(value)

if 'offset_by' in answer_min_or_max:
    offset = answer_min_or_max['offset_by']
    value += relativedelta(years=offset.get('years', 0),
                           months=offset.get('months', 0),
                           days=offset.get('days', 0))
```

nealedj commented on app/forms/date_form.py:
```diff
@@ -84,3 +98,62 @@ class MonthYearDateForm(Form):
+
+    if 'offset_by' in answer_min_or_max:
+        offset = answer_min_or_max['offset_by']
+        value = convert_to_datetime(value) + relativedelta(years=offset.get('years', 0),
```
Is there any reason that `relativedelta` is being used rather than the basic, builtin `datetime.timedelta`?

shahi645 replied:
relativedelta deals with months and year and timedelta only days

nealedj commented on app/forms/questionnaire_form.py:
```diff
@@ -105,6 +111,16 @@ def _validate_calculated_question(self, calculation, question, target_total, cur
         return True
 
+    @staticmethod
+    def _get_period_limits(limits):
```
Could be simplified to:

```
limits = limits or {}
return limits.get('minimum'), limits.get('maximum')
````

nealedj commented on app/validation/error_messages.py:
```diff
@@ -22,4 +22,8 @@
     'MAX_LENGTH_EXCEEDED': 'Your answer is too long, it has to be less than %(max)d characters.',
     'INVALID_DATE': 'Enter a valid date.',
     'INVALID_DATE_RANGE': "Enter a 'period to' date later than the 'period from' date.",
+    'DATE_PERIOD_TOO_SMALL': 'Enter a reporting period greater than %(min)s.',
```
I'm sure this conversation has already happened - but why are we hard-coding a specific error message about reporting periods when the validation is generic date range validation?

shahi645 replied:
Speaking to Will he just said to use this for now, and then we can change as we need

nealedj commented on app/validation/validators.py:
```diff
@@ -73,9 +75,10 @@ def __init__(self, minimum=None, minimum_exclusive=False,
-        if not messages:
-            messages = error_messages
-        self.messages = messages
+        if messages:
```
Could be simplified to:

`self.messages = messages or error_messages`

shahi645 replied:
nice

nealedj commented on app/validation/validators.py:
```diff
@@ -212,31 +216,93 @@ def __call__(self, form, field):
+        if self.minimum_date:
+            if date < self.minimum_date:
+                raise validators.ValidationError(self.messages['SINGLE_DATE_PERIOD_TOO_EARLY'] %
+                                                 dict(min=self._format_playback_date(self.minimum_date + relativedelta(days=-1))))
```
Same comment again about `relativedelta` - is it giving us anything that we don't already have with `datetime.timedelta`?

shahi645 replied:
answered above, actually looking at it again, it only deals with days, second, microseconds weirdly 😞

nealedj commented on app/validation/validators.py:
```diff
@@ -283,3 +349,10 @@ def format_playback_value(value, currency=None):
+
+
+def format_date_string(field):
+    day = int(field.data['day']) if 'day' in field.data else 1
```
Simplified:

`day = int(field.data.get('day', 1))`

nealedj reviewed (changes requested):
Looks good - nearly all my comments are minor.

The only big thing might be whether this validation will work within a repeating group.

dwyeradam commented on app/forms/fields.py:
```diff
@@ -15,16 +15,17 @@
             'Radio': get_select_field,
             'Checkbox': get_select_multiple_field,
-            'Date': get_date_field,
             'MonthYearDate': get_month_year_field,
```
Does `MonthYearDate` not also need to allow for date validation? Please don't say this isn't a future requirement unless you tell me the winning numbers for the next lottery draw.

shahi645 replied:
I don't think so as they work on a referenced date, so wouldn't be required

dwyeradam replied:
I'm pretty sure the LFS stuff @davecharles was working on had Month Year dates for when you arrived in the UK that couldn't be before dob. We may not need date range but are we sure we don't need single validation setting min and max? I could see a future need there but @nealedj what are your thoughts? @charlotte-hirst are you expecting this for all date types?

shahi645 replied:
Can you advise @charlotte-hirst also. As you are in contact with Social & Census

nealedj replied:
I think we us to consider date range validation "done" then we should support all types of date question. However, I don't think there's any reason to block this PR from going in at the moment. The code changes are becoming pretty large and it's becoming exponentially more difficult to review.

I'd be in favour of closing this off as it adds much-needed value to master. The job of applying the validation to `MonthYearDate` could be considered a separate piece of work - or Part 2 (the validation strikes back).

Let's keep this sort of thing in check though. In an ideal world this PR would cover all use-cases.

shahi645 replied:
OK will put a ticket into backlog with some fleshed out AC

dwyeradam commented on app/forms/questionnaire_form.py:
```diff
@@ -66,10 +85,40 @@ def _get_target_total_and_currency(self, calculation, question):
         return self.answer_store.filter(answer_ids=[target_answer['id']]).values()[0], target_answer.get('currency')
 
-    def _validate_date_range_question(self, question_id, period_from_id, period_to_id):
+    def validate_date_range_with_period_limits_and_single_date_limits(self, question_id, period_limits, date_from, date_to):  # pylint: disable=too-many-locals
```
Remove `# pylint: disable=too-many-locals`. It will pass without this. If it didn't pass Lint you should really address the code rather than add an ignore.

dwyeradam commented on app/forms/questionnaire_form.py:
```diff
@@ -66,10 +85,40 @@ def _get_target_total_and_currency(self, calculation, question):
 
-    def _validate_date_range_question(self, question_id, period_from_id, period_to_id):
+    def validate_date_range_with_period_limits_and_single_date_limits(self, question_id, period_limits, date_from, date_to):  # pylint: disable=too-many-locals
+        if self._has_min_and_max_single_dates(date_from, date_to):
```
This would be clearer if it was after [`if question['type'] == 'DateRange':`](https://github.com/ONSdigital/eq-survey-runner/pull/1521/files?diff=unified#diff-c5e43bea5b19fce9acef957e0a446c22R44) as both the methods that are then called below start with that check. You can then also remove it from  [`validate_date_range_with_single_date_limits`](https://github.com/ONSdigital/eq-survey-runner/pull/1521/files?diff=unified#diff-c5e43bea5b19fce9acef957e0a446c22R106). 
It's better practice to only call a method when needed and not to call it regardless and let the method decide if it should have been called.

dwyeradam reviewed (changes requested):
A couple of little early points while I continue to review.

dwyeradam commented on app/forms/questionnaire_form.py:
```diff
@@ -37,10 +42,26 @@ def validate(self):
-                if not (self.answers_all_valid([period_from_id, period_to_id]) and
-                        self._validate_date_range_question(question['id'], period_from_id, period_to_id)):
+                date_from = question['answers'][0]
+                date_to = question['answers'][1]
```
There is quite a bit of code in this if statement now. It may be worth considering moving it into a `self.validate_daterange_question(.....)` method as it does below for calculation and returning the result.

dwyeradam commented on app/forms/questionnaire_form.py:
```diff
@@ -37,10 +42,26 @@ def validate(self):
+                                                           question.get('period_limits'))):
                     valid_form = False
             elif question['type'] == 'Calculated':
                 valid_form = self.validate_calculated_question(question)
```
You haven't added this issue in this PR but applying "Boy Scout" rule. If we ever got a block with a date range and a calculation on it (we shouldn't do but I don't think there is anything to stop that possibility) the calculation validation will override the setting of valid_form.

nealedj replied:
Good catch.

dwyeradam reviewed (changes requested):
Two little things but looking good. Only real issue we still need to discuss is the fact that no validation is available for MonthYear dates.

MIslam92 reviewed (approved):
LGTM

MebinAbraham commented on app/validation/error_messages.py:
```diff
@@ -22,4 +22,8 @@
     'INVALID_DATE': 'Enter a valid date.',
     'INVALID_DATE_RANGE': "Enter a 'period to' date later than the 'period from' date.",
+    'DATE_PERIOD_TOO_SMALL': 'Enter a reporting period greater than %(min)s.',
+    'DATE_PERIOD_TOO_LARGE': 'Enter a reporting period less than %(max)s.',
```
The error messages are wrong.
```
"period_limits": {
	"minimum": {
		"days": 31
     },
	"maximum": {
		"days": 31
	}
}
```
If you use the above settings and if the period difference is 32, the error message shown is:
`1) Enter a reporting period less than 31 days.`
It should be: `less than 32 days` or `less than or equal to 31 days`

If the period difference is 30, the error message shown is:
`1) Enter a reporting period greater than 31 days.`
It should be: `greater than 30 days` or `greater than or equal to 31 days`

The range itself is fine, just need to tweak the error messages. This will apply to `days`, `months` and `years`.

shahi645 replied:
Schema validator wont allow same period limits but error messages are a bit strange there i will dig around a little and get back

shahi645 replied:
Just looking at these again the error messages are correct

MebinAbraham replied:
Minimum and maximum are generally inclusive. But also your point about schema validator not allowing?. For me, it says it is a valid schema?

Also, what am I missing?. How is it a valid error message?. `31` is a valid range but the error message would say ie less than `31`.?

dwyeradam reviewed (changes requested):
On `test_single_date_period_validation.json` if you don't enter anything on the second page it throws a 500 error. If it's not Mandatory we wouldn't want to 500 although in this instance it still does. Edge case but at minimum a new card. The issue is that no valid range can be entered, not entering anything however would be valid.

MebinAbraham reviewed (approved):
The error messages has been sorted.
I think Adam point can be raised as a new card and this can go in.
There are certain edge cases we can improve but for now, looks good 👍 
I think the other cards raised from this PR should be moved to the top of ready for dev, to be picked up after MBS goes in.

dwyeradam reviewed (approved):
I relent, lets get this bad boy in. 
It would help to add some guidance into `test_single_date_period_validation.json` to help guide what date to enter that won't 500.

dwyeradam commented:
@shahi645, squishy, squashy, good commit messagey.

shahi645 commented:
@dwyeradam Ill add the guidance, thats a good point. Update with aster. Then squash

shahi645 committed:
Added date validation at question and answer level

question level:

- DateRange question type now has validation for period entered

- period can have a minimum or a maxmimum or both

- period can be set in days, months, years or any combination

- runtime validation added for when a schema is invalid when there is answer level validation also

answer level:

- Date answer type can now have minimum or maximum

- same pattern used for date_comparisons in routing rules

- offset can be set day, months, years or any combination

- offset is from a referenced date, which is either a value, meta or a previous answer id

- runtime validation added for invalid schemas, i.e. minimum greter than maximum and when question validation also in place
